lib/load-config.js

```diff
@@ -22,30 +22,131 @@ const CONFIG_FILE_NAMES = Object.freeze([
+class UnsupportedFileTypeError extends Error {
+  constructor (path) {
+    super(
+      `Unsupported file type ${extname(
+        path
+      ).toLowerCase()} while reading file ${path}`
+    )
+    this.name = UnsupportedFileTypeError.name
+  }
+}
+
+class ConfigParsingError extends Error {
+  constructor (path, errorDetails, originalError) {
+    super(
+      `Error loading configuration from file ${path}: ${errorDetails}` +
+        (typeof originalError === 'object' && originalError instanceof Error
+          ? ` Original error: ${originalError.message}`
+          : '')
+    )
+
+    this.name = UnsupportedFileTypeError.name
+
+    if (originalError instanceof Error) {
+      this.stack = originalError.stack
+    }
   }
+}
+
+const NO_EXPORTS = Symbol('no exports')
+
+function loadConfigFile (path, _di) {
+  const di = {
+    readFile: (path) => readFileSync(path, 'utf8'),
+    readJs: (path) => require(path),
+    ..._di
+  }
+
+  let config
+
+  const fileName = basename(path)
+  const ext = extname(path).toLowerCase()
+
+  if (EXTS_YAML_VALIDATOR.test(ext)) {
+    const JsYaml = require('js-yaml')
+    try {
+      // TODO: add YAML schema so that we get better errors for YAML users.
+      config = JsYaml.load(di.readFile(path))
+    } catch (error) {
+      if (typeof error === 'object') {
+        if (error instanceof JsYaml.YAMLException) {
+          throw new ConfigParsingError(
+            path,
+            'must contain a valid c8 configuration object.',
+            error
+          )
+        }
+      }
+
+      throw error
+    }
+
+    if (!config) {
+      // TODO: remove this check once we get YAML schema validation.
+      throw new ConfigParsingError(path, 'invalid configuration')
+    }
+  } else {
+    if (EXTS_JS_VALIDATOR.test(ext)) {
+      // Add a loader that allows us to check to see if anything was ever exported.
+      // Thank you to https://stackoverflow.com/a/70999950 for the inspiration. Please note that this code won't port to TypeScript nicely.
+      const extensions = module.constructor._extensions
+      const cjsLoader = extensions['.cjs']
+      const jsLoader = extensions['.js']
+      extensions['.cjs'] = extensions['.js'] = (module, filename) => {
+        module.exports[NO_EXPORTS] = filename
+        jsLoader(module, filename)
+      }
+      try {
+        config = di.readJs(path)
+      } finally {
+        // Undo the global state mutation, even if an error was thrown.
+        extensions['.cjs'] = cjsLoader
+        extensions['.js'] = jsLoader
+      }
+      if (NO_EXPORTS in config) {
+        throw new ConfigParsingError(
+          path,
+          'does not export a c8 configuration object.'
+        )
+      }
+    } else if (ext === '.json' || fileName.slice(-2) === 'rc') {
+      try {
+        config = JSON.parse(di.readFile(path))
+      } catch (error) {
+        if (typeof error === 'object') {
+          if (error instanceof SyntaxError) {
```

Comment:
Got rid of the unnecessary check.  You can check on the type of any variable with `instanceof`.

[Addressed](https://gist.github.com/mcknasty/80c8c814864241de05987dc3cae5fd54#file-parse-args-patch-diff-L207-L208) in the review branch

Reply:
See previous response on this subject.